### PR TITLE
 Persistent storage dashboard : Added changes to set correct ceph cluster name in health card

### PR DESCRIPTION
**Pull request by anmolsachan**
Signed-off-by: Anmol Sachan <anmol13694@gmail.com>

**When both cluster name and cluster health status are available:**
Screenshot from 2019-08-21 00-25-42

**When cluster name is available but health status is not available:**
Screenshot from 2019-08-21 00-20-56

**When both cluster name and health status are unavailable:**
Screenshot from 2019-08-21 00-20-37

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/health-card.tsx [hunk 1]
```diff
@@ -12,12 +14,25 @@ import {
...
 import { HealthState } from '@console/internal/components/dashboard/health-card/states';
+import { K8sResourceKind, referenceForModel } from '@console/internal/module/k8s';
+import { CephClusterModel } from '../../../../models';
+import { CEPH_STORAGE_NAMESPACE } from '../../../../constants/index';
 import { STORAGE_HEALTH_QUERIES, StorageDashboardQuery } from '../../../../constants/queries';
 import { filterCephAlerts } from '../../../../selectors';
-import { getCephHealthState } from './utils';
+import { getCephHealthState, processCephHealthState } from './utils';
+
+
```
Comment (gnehapk) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/health-card.tsx]: Nit: Please remove the extra line here

**Comment by rawagner** [comment]
please remember that you also need to show correct name in Overview dashboard health card. Storage health check is done via this extension point https://github.com/openshift/console/blob/master/frontend/packages/ceph-storage-plugin/src/plugin.ts#L146

**Comment by cloudbehl** [comment]
@anmolsachan Please add the screenshots of various conditions we are handling.  The Lint errors are failing please fix them

**Comment by anmolsachan** [comment]
> please remember that you also need to show correct name in Overview dashboard health card. Storage health check is done via this extension point https://github.com/openshift/console/blob/master/frontend/packages/ceph-storage-plugin/src/plugin.ts#L146

@rawagner I agree, but there is a problem to that. The `healthHandler: getCephHealthState` expects only one argument to be passed which for the subsystem health. Now in the case of ceph, we need the cluster name along with it. 
What should be the right approach to handle this issue ?
One approach I see is : watch for a subsystem health (ceph here) also in (https://github.com/openshift/console/blob/master/frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx ) and pass it to the healthhandler and then it processes it and gives it back to display. 
Can you please suggest something here ?

**Comment by rawagner** [comment]
We would need to add another prop here https://github.com/openshift/console/blob/master/frontend/packages/console-plugin-sdk/src/typings/dashboards.ts#L37 (ie resource) which would specify if there's any resource to fetch additionally to the prometheus query. The resource would then be passed to `healthHandler`. I can work on that part if you prefer.

**Comment by anmolsachan** [comment]
> We would need to add another prop here https://github.com/openshift/console/blob/master/frontend/packages/console-plugin-sdk/src/typings/dashboards.ts#L37 (ie resource) which would specify if there's any resource to fetch additionally to the prometheus query. The resource would then be passed to `healthHandler`. I can work on that part if you prefer.

@rawagner That would be great if you could do that. I wouldn't want to break things doing it, since I don't know the complete flow about it .

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts [hunk 2]
```diff
@@ -29,6 +30,17 @@ export const getCephHealthState = (ocsResponse): CephHealth => {
   return CephHealthStatus[value] || CephHealthStatus[3];
 };
 
+export const processCephHealthState = (cephHealthState, cephClusterData): CephHealth => {
+  const cephClusterName = getName(_.get(cephClusterData, 0));
```
Comment (gnehapk) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts]: add types for cephHealthState, cephClusterData

**Comment by anmolsachan** [comment]
PR dependent on https://github.com/openshift/console/pull/2480. Thanks to @rawagner for the quick resolution and help.

Ceph health status with name now available on Overview dashboard as well :
Screenshot from 2019-08-27 18-46-23

File: frontend/packages/ceph-storage-plugin/src/constants/resources.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { FirehoseResource } from '@console/internal/components/utils/index';
+import { referenceForModel } from '@console/internal/module/k8s/k8s';
+import { CephClusterModel } from '../models';
+import { CEPH_STORAGE_NAMESPACE } from './index';
+
+const cephClusterResource: FirehoseResource = {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/constants/resources.ts]: should be synced with https://github.com/openshift/console/pull/2433/ - ideally reuse the same definition

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts [hunk 4]
```diff
@@ -21,12 +23,18 @@ const CephHealthStatus = [
   },
 ];
 
-export const getCephHealthState = (ocsResponse): CephHealth => {
+export const getCephHealthState = (ocsResponse, cephCluster): CephHealth => {
   if (!ocsResponse) {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts]: you should also wait for `cephCluster` loaded/error state

**Comment by anmolsachan** [comment]
@rawagner Can you please review this, I have incorporated the review comments. Thanks

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/health-card.tsx [hunk 5]
```diff
@@ -15,33 +16,59 @@ import {
...
 import { HealthState } from '@console/internal/components/dashboard/health-card/states';
+import { referenceForModel } from '@console/internal/module/k8s';
+import { FirehoseResource } from '@console/internal/components/utils';
+import { CephClusterModel } from '../../../../models';
+import { CEPH_STORAGE_NAMESPACE } from '../../../../constants/index';
 import { STORAGE_HEALTH_QUERIES, StorageDashboardQuery } from '../../../../constants/queries';
 import { filterCephAlerts } from '../../../../selectors';
 import { getCephHealthState } from './utils';
 
+const cephClusterResource: FirehoseResource = {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/health-card.tsx]: this is still old definition. Can you specify the definition in one place and reuse it everywhere so we will avoid these issues in the future ?

Comment (anmolsachan) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/health-card.tsx]: Have defined the definition at frontend/packages/ceph-storage-plugin/src/constants/resources.ts 
Forgot to change it here. Thanks for the catch. Will update

**Comment by anmolsachan** [comment]
@rawagner I have made the change. Please review.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts [hunk 6]
```diff
@@ -21,12 +23,18 @@ const CephHealthStatus = [
   },
 ];
 
-export const getCephHealthState = (ocsResponse): CephHealth => {
-  if (!ocsResponse) {
+export const getCephHealthState = (ocsResponse, cephCluster): CephHealth => {
+  if (!ocsResponse || !cephCluster) {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/health-card/utils.ts]:
```suggestion
  if (!ocsResponse || !_.get(cephCluster, 'loaded')) {
```

other than that, lgtm. Once https://github.com/openshift/console/pull/2480 is merged, please rebase and we can merge

**Comment by anmolsachan** [comment]
@rawagner @cloudbehl I have made the requested changes and rebased on top of master, please review.

**Pull request merged**